File: apps/desktop/src/platform/storage.ts
// apps/desktop/src/platform/storage.ts
//
// (, , P-NEW-1) — real StoragePort backed by
// @tauri-apps/plugin-store, with durable writes routed through the Rust
// `metadata_save_atomic` command (P-NEW-1 fix).
//
// Why not just call store.save()?
// tauri-plugin-store@2.4.2 implements save() as fs::write — non-atomic.
// demands "old-or-new on power loss"; only an atomic temp-file +
// rename gives that guarantee. The Rust `metadata_save_atomic` command
// in src-tauri/src/storage.rs uses `tempfile::NamedTempFile` + `persist`
// which is `std::fs::rename` — atomic on POSIX & NTFS.
//
// Persistence contract preserved across mobile/desktop:
// Single store file `wallet-state.json` at BaseDirectory::AppData.
// One top-level key per Zustand store (`prl_wallet_registry`,
// `prl-pin`, etc.). Wallet metadata (wallets[], activeWalletId) lands
// under `prl_wallet_registry`; walletType is NEVER serialized into
// this file ( invariant).
//
// Lazy load: the Store handle is constructed once per process (first
// getItem/setItem/removeItem call) and cached. `autoSave: false` means we
// own the durable-write trigger explicitly — exactly one
// `metadata_save_atomic` per setItem / removeItem.
//
// Path resolution: @tauri-apps/plugin-store@2.4.2 does not expose
// Store.path() on its public API (verified against dist-js/index.d.ts).
// We compute the on-disk path the same way the plugin does:
// appDataDir() / STORE_FILE
// and pass it to the Rust atomic-save command.

import { invoke } from "@tauri-apps/api/core";
import { appDataDir, join } from "@tauri-apps/api/path";
import { load, type Store } from "@tauri-apps/plugin-store";
import type { StoragePort } from "@prl-wallet/app-adapters";

const STORE_FILE = "wallet-state.json";

let storePromise: Promise<Store> | null = null;
let pathPromise: Promise<string> | null = null;

function getStore(): Promise<Store> {
  if (!storePromise) {
    storePromise = load(STORE_FILE, { autoSave: false, defaults: {} });
  }
  return storePromise;
}

function getStorePath(): Promise<string> {
  if (!pathPromise) {
    pathPromise = (async () => {
      const dir = await appDataDir();
      return join(dir, STORE_FILE);
    })();
  }
  return pathPromise;
}

/**
 * Build the JSON payload that metadata_save_atomic writes. The shape
 * matches what tauri-plugin-store's Store.save() would write (an object
 * keyed by every store entry), so a Wave 4 SIGKILL UAT can interleave
 * Store.save() and metadata_save_atomic without payload divergence.
 */
async function serializeStoreState(store: Store): Promise<string> {
  const entries = await store.entries();
  const obj: Record<string, unknown> = {};
  for (const [k, v] of entries) {
    obj[k] = v;
  }
  return JSON.stringify(obj);
}

async function durableWrite(store: Store): Promise<void> {
  const path = await getStorePath();
  const contents = await serializeStoreState(store);
  // P-NEW-1 fix: route the durable write through the Rust atomic command,
  // NOT store.save() (which is non-atomic fs::write).
  await invoke<void>("metadata_save_atomic", {
    path,
    contents,
  });
}

export function createDesktopStorage(): StoragePort {
  return {
    async getItem(key) {
      const store = await getStore();
      const value = await store.get<string>(key);
      return value ?? null;
    },
    async setItem(key, value) {
      const store = await getStore();
      await store.set(key, value);
      await durableWrite(store);
    },
    async removeItem(key) {
      const store = await getStore();
      await store.delete(key);
      await durableWrite(store);
    },
  };
}

File: apps/desktop/src-tauri/src/storage.rs
// apps/desktop/src-tauri/src/storage.rs
//
// (P-NEW-1, , ) — atomic metadata write command.
//
// `tauri-plugin-store::Store::save()` writes via fs::write (non-atomic).
// demands "old-or-new on power loss"; only an atomic temp-file +
// rename gives that guarantee. This module owns that primitive.
//
// JS-side StoragePort (apps/desktop/src/platform/storage.ts) lands in Wave 2
// and routes setItem/removeItem durable writes through `metadata_save_atomic`
// instead of (or in addition to) Store.save().
//
// Dev-only `__test_write_metadata` is the SIGKILL UAT trigger (). The
// command vanishes from release builds via cfg(debug_assertions), and the
// handler entry in lib.rs is correspondingly cfg-gated — even a compromised
// frontend that knows the command name gets "command not found" in release.

use std::fs;
use std::io::Write;
use std::path::PathBuf;
use tempfile::NamedTempFile;

/// Atomic write helper.
/// `path`: absolute file path the JS adapter computed via `Store.path()` /
/// `BaseDirectory.AppData`. Rust does not compute the path itself
/// to keep the directory-resolution logic in JS land where Wave 2's
/// `@tauri-apps/plugin-store` lives.
/// `contents`: the full serialized JSON the JS adapter computed (Store.set
/// followed by Store.save with autoSave:false leaves the in-memory
/// state correct; Wave 2 reads back via Store.entries() then
/// JSON.stringify-s into `contents`).
fn atomic_write(path: &str, contents: &str) -> Result<(), String> {
    let target = PathBuf::from(path);
    let parent = target
        .parent()
        .ok_or_else(|| "metadata_save_atomic: path has no parent dir".to_string())?;
    fs::create_dir_all(parent).map_err(|e| format!("create_dir_all: {e}"))?;
    let mut tmp =
        NamedTempFile::new_in(parent).map_err(|e| format!("NamedTempFile::new_in: {e}"))?;
    tmp.write_all(contents.as_bytes())
        .map_err(|e| format!("write_all: {e}"))?;
    tmp.as_file()
        .sync_data()
        .map_err(|e| format!("sync_data: {e}"))?;
    // tempfile::persist performs std::fs::rename — atomic on POSIX & NTFS.
    tmp.persist(&target)
        .map_err(|e| format!("persist (rename): {}", e.error))?;
    Ok(())
}

/// Production atomic-save command. Wave 2's StoragePort.setItem and
/// StoragePort.removeItem call this after Store.set/Store.delete to
/// satisfy .
#[tauri::command]
pub fn metadata_save_atomic(path: String, contents: String) -> Result<(), String> {
    atomic_write(&path, &contents)
}

/// Dev-only mid-write trigger for the SIGKILL UAT ().
/// Compile-time stripped from release binaries; the handler entry in lib.rs
/// is also cfg-gated. UAT script (Wave 4 — scripts/uat/atomicity.sh) invokes
/// this via a debug-only CLI flag on the binary.
///
/// Behavior is identical to `metadata_save_atomic` so the SIGKILL test
/// exercises the exact same write path; the only difference is the
/// compile-time gate.
#[cfg(debug_assertions)]
#[tauri::command]
pub fn __test_write_metadata(path: String, contents: String) -> Result<(), String> {
    atomic_write(&path, &contents)
}
